crates/store/src/db/sql.rs

```diff
@@ -184,41 +189,125 @@ pub fn select_accounts_by_ids(
+        Err(e) => return Err(e.into()),
+    };
+
+    #[derive(num_derive::FromPrimitive)]
+    enum RecordType {
+        StorageScalars = 0,
+        StorageMapValues,
+        FungibleAssets,
+        NonFungibleAssets,
+    }
+    let mut select_merged_deltas_stmt =
+        conn.prepare_cached(include_str!("sql-queries/select_merged_deltas.sql"))?;
+    let mut rows = select_merged_deltas_stmt.query(params![account_id, block_start, block_end])?;
+
+    enum FieldIndex {
+        RecordType = 0,
+        Slot = 2,
+        Key = 3,
+        Value = 4,
+    }
+
+    let mut storage_scalars = BTreeMap::new();
+    let mut storage_maps = BTreeMap::new();
+    let mut fungible = BTreeMap::new();
+    let mut non_fungible_delta = NonFungibleAssetDelta::default();
     while let Some(row) = rows.next()? {
-        let delta = AccountDelta::read_from_bytes(row.get_ref(0)?.as_blob()?)?;
-        result.push(delta);
+        let record_type = RecordType::from_usize(row.get(FieldIndex::RecordType as usize)?)
+            .expect("Record type value must be one of the `RecordType` enum variants");
+        match record_type {
+            RecordType::StorageScalars => {
+                let slot = row.get(FieldIndex::Slot as usize)?;
+                let value_data = row.get_ref(FieldIndex::Value as usize)?.as_blob()?;
+                let value = Word::read_from_bytes(value_data)?;
+                storage_scalars.insert(slot, value);
+            },
+            RecordType::StorageMapValues => {
+                let slot = row.get(FieldIndex::Slot as usize)?;
+                let key_data = row.get_ref(FieldIndex::Key as usize)?.as_blob()?;
+                let key = Digest::read_from_bytes(key_data)?;
+                let value_data = row.get_ref(FieldIndex::Value as usize)?.as_blob()?;
+                let value = Word::read_from_bytes(value_data)?;
+
+                match storage_maps.entry(slot) {
+                    Entry::Vacant(entry) => {
+                        entry.insert(StorageMapDelta::new(BTreeMap::from([(key, value)])));
+                    },
+                    Entry::Occupied(mut entry) => {
+                        entry.get_mut().insert(key, value);
+                    },
+                }
+            },
+            RecordType::FungibleAssets => {
+                let faucet_id: u64 = row.get(FieldIndex::Key as usize)?;
+                let value = row.get(FieldIndex::Value as usize)?;
+
+                fungible.insert(faucet_id.try_into()?, value);
+            },
+            RecordType::NonFungibleAssets => {
+                let vault_key_data = row.get_ref(FieldIndex::Key as usize)?.as_blob()?;
+                let vault_key = Word::read_from_bytes(vault_key_data)?;
+
+                // SAFETY: This conversion is safe because we previously wrote this key from
+                //   internal value of `NonFungibleAsset`.
+                let asset = unsafe { NonFungibleAsset::new_unchecked(vault_key) };
```

Comment:
Note that this could occur in practice due to disk corruption. I'm not sure we should allow `unsafe` here.

Reply:
I missed that we have another way for `NonFungibleAsset` construction, thank you for noticing!